File: typescript/src/problems1_100/jump-game2.ts
class Problem {
  jump(nums: number[]): number {
    throw new Error(`Not implemented (${nums}).`);
  }
}

class Default extends Problem {
  jump(nums: number[]): number {
    let steps: number = 0;
    let maxLastJump: number = 0;
    let maxCurrentJump: number = 0;
    for (let i = 0; i < nums.length - 1; i++) {
      maxCurrentJump = Math.max(maxCurrentJump, i + nums[i]);
      if (i !== maxLastJump) {
        continue;
      }
      steps++;
      maxLastJump = maxCurrentJump;
    }
    return steps;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems1_100/jump_game2.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def jump(self, nums: list[int]) -> int:
        pass


class Default(Problem):
    @override
    def jump(self, nums: list[int]) -> int:
        steps: int = 0
        max_last_jump: int = 0
        max_current_jump: int = 0
        for i in range(len(nums) - 1):
            max_current_jump = max(max_current_jump, i + nums[i])
            if i != max_last_jump:
                continue
            steps += 1
            max_last_jump = max_current_jump
        return steps


SOLUTIONS: list[Problem] = [Default()]
